src/gossip.rs

```diff
@@ -0,0 +1,235 @@
+pub(crate) const GOSSIP_SYNC_TIME_KEY: &str = "last_sync_timestamp";
+pub(crate) const NETWORK_GRAPH_KEY: &str = "network_graph";
+
+struct Gossip {
+    pub last_sync_timestamp: u32,
+    pub network_graph: NetworkGraph,
+}
+
+impl Gossip {
+    pub fn new(network: Network, logger: Arc<MutinyLogger>) -> Self {
+        Self {
+            last_sync_timestamp: 0,
+            network_graph: NetworkGraph::new(network, logger),
+        }
+    }
+}
+
+async fn build_gossip_database() -> Result<Rexie, MutinyError> {
+    // Create a new database
+    let rexie = Rexie::builder(GOSSIP_DATABASE_NAME)
+        // Set the version of the database to 1.0
+        .version(1)
+        .add_object_store(ObjectStore::new(GOSSIP_OBJECT_STORE_NAME))
+        // Build the database
+        .build()
+        .await?;
+
+    Ok(rexie)
+}
+
+async fn get_gossip_data(rexie: &Rexie, logger: Arc<MutinyLogger>) -> Result<Gossip, MutinyError> {
+    // Create a new read-only transaction
+    let transaction = rexie.transaction(&[GOSSIP_OBJECT_STORE_NAME], TransactionMode::ReadOnly)?;
+
+    let store = transaction.store(GOSSIP_OBJECT_STORE_NAME)?;
+
+    // Get the `last_sync_timestamp`
+    let last_sync_timestamp_js = store.get(&JsValue::from(GOSSIP_SYNC_TIME_KEY)).await?;
+    let last_sync_timestamp: u32 =
+        last_sync_timestamp_js
+            .as_f64()
+            .ok_or_else(|| MutinyError::ReadError {
+                source: MutinyStorageError::Other(anyhow!(
+                    "could not read last_sync_timestamp, got {:?}",
+                    last_sync_timestamp_js
+                )),
+            })? as u32;
+
+    // Get the `network_graph`
+    let network_graph_js = store.get(&JsValue::from(NETWORK_GRAPH_KEY)).await?;
+
+    let network_graph_str: String = serde_wasm_bindgen::from_value(network_graph_js)?;
+    let network_graph_bytes: Vec<u8> = Vec::from_hex(&network_graph_str)?;
+    let mut readable_bytes = lightning::io::Cursor::new(network_graph_bytes);
+    let network_graph = NetworkGraph::read(&mut readable_bytes, logger)?;
+
+    let gossip = Gossip {
+        last_sync_timestamp,
+        network_graph,
+    };
+
+    Ok(gossip)
+}
+
+async fn write_network_graph(
+    rexie: &Rexie,
+    last_sync_timestamp: u32,
+    network_graph: &NetworkGraph,
+) -> Result<(), MutinyError> {
+    // Create a new read-write transaction
+    let transaction = rexie.transaction(&[GOSSIP_OBJECT_STORE_NAME], TransactionMode::ReadWrite)?;
+
+    let store = transaction.store(GOSSIP_OBJECT_STORE_NAME)?;
+
+    // Save the last sync timestamp
+    store
+        .put(
+            &JsValue::from(last_sync_timestamp),
+            Some(&JsValue::from(GOSSIP_SYNC_TIME_KEY)),
+        )
+        .await?;
+
+    // Save the network graph
+    let network_graph_js = serde_wasm_bindgen::to_value(&network_graph.encode().to_hex())?;
+    store
+        .put(&network_graph_js, Some(&JsValue::from(NETWORK_GRAPH_KEY)))
+        .await?;
```

Comment:
Should we delete the `GOSSIP_SYNC_TIME_KEY` we just saved? Or reset it back to the last value? or does `transaction.done()` make sure that both happen?

Reply:
The transactions are supposed to be atomic so that shouldn't be a concern